packages/core/lib/console.js

```diff
@@ -61,13 +61,16 @@ class Console extends EventEmitter {
 
   async start() {
     try {
+      const context = {};
+      this.repl = { context };
+      await this.setUpEnvironment();
+      this.provision();
+
```

Comment:
We should document this. This is necessary because we want the repl's initial context to be provisioned and set up appropriately because `repl.start()` will immediately start consuming `process.stdin`, which would makes testing difficult. This change ensures that repl's env is provisioned correctly in normal and test use cases.

Maybe it's worth factoring out these lines to a helper function `getInitialREPLContext()`?

Reply:
So what is really going on here is that we're shortening the time from when the repl starts and the context is added. This way all values are calculated ahead of time and injected directly after. I almost made a comment and decided not to, I'll add one!

Reply:
It seems that whenever a node repl is created (`repl.start(options);` or `new repl.REPLServer(options);`) it automatically starts consuming its `stdin`. That is, there isn't a way to start a repl server without it automatically starting to consume `stdin`. My hypothesis is this the time between consuming stdin and resolving its context is not a problem when the repl is connected to a terminal (normal usage, invoking the command line) but when it starts as an exec'd child-process `stdin` is immediately consumed (before resolving the context).

The other issue here is the `context = {}; this.repl = { context }` business which looks odd but is necessary because of how repl creation works. The functions to update the contexts mutate `this.repl.context` and this is the current solution. :/ 

I'm still not happy with it, but I'd like a comment to remind the next person of the reason for the `context = {}; this.repl = { context }` business

Reply:
Why wouldn't they just let you input the context as an argument to `start`?

Reply:
I'm going to just edit `setUpEnvironment` to return the context instead of doing the mutating - why not?

Reply:
It really is odd! The only pattern I see in the docs is to use a fluent style to set the context immediately after invoking `repl.start()`
```
const repl = require('repl');
const msg = 'message';

repl.start('> ').context.m = msg;
```